### ISSUE TITLE
Using built-in tag via JSX literal incorrectly counts as variable usage

### ISSUE THREAD

**Issue by glenjamin** [eslint/eslint#1911]
Minimal reproducing scenario:
```jsx
var div = 1;
React.render(<div />);
```
I would expect this to warn on unused `div` variable, but it does not.

React is in the process of removing the tag whitelist in JSX - it's slated to land in React 0.13 (facebook/react#2746).

The current rule is that JSX tags beginning with a lowercase character will transform to a string, and others will transform to a variable. I guess it's an open question whether this can be considered a part of the "JSX spec", or a react-specific implementation details

This logic to special-case tags that start with a lowercase character already exists in the no-undef rule, so the suggested fix is just to port this over to no-unused-vars.

**Comment by lo1tuma** [comment]
Seems like this special case is specific to react. Maybe we should remove this special case logic from `no-undef` or at least make it an opt-in option.

**Comment by glenjamin** [comment]
It's tricky to say, the only other JSX compilers I could find were https://www.npmjs.com/package/virtual-dom-jsx and https://github.com/insin/msx

Their examples all show using lowercase strings, and from reading their implementations they appear to follow React's rules about strings vs variables.

**Comment by nzakas** [comment]
This is definitely a bug as it behaves today. The real question is the solution. 

I could see an argument for removing the current exceptions altogether, as this seems to be a React - specific set of semantics (which is why we are in this situation in the first place). But to remove it, we would need to expose something so plugins could implement this same functionality.

**Comment by glenjamin** [comment]
The decision to support JSX at all was a pragmatic one, I think treating React's approach to variable vs strings for "tag names" as the official approach is in-line with that choice (while supporting the `React.createElement` stuff is a step too far as already discussed).

I suspect if the existing JSX semantic knowledge is removed from these rules it will generate quite a lot more queries from users :(

**Comment by nzakas** [comment]
I'm not sure if it was pragmatic or just misinformed. The fact that semantics have changed means we likely made the wrong choice. We can't be caught in a situation where the core rules are at the mercy of a library defining what the semantics of some syntax is.

So I think the right way forward here is to remove the exception and add a way for doors to mark a variable as used. Then we can work with @yannickcr to make sure eslint-plugin-react is updated to make up for this change.

**Comment by glenjamin** [comment]
Would that just be a case of making https://github.com/eslint/eslint/blob/7bb738f67fbe441e8769134ede1a46431614c1dc/lib/rules/no-unused-vars.js#L102 an officially documented thing?

I should be able to extend the stuff i did in yannickcr/eslint-plugin-react#6 to make that work.

Would you also want to remove the JSX variable support from `no-undef` ?

**Comment by yannickcr** [comment]
I am open to update eslint-plugin-react to handle these cases, but, as you said, we need a way to mark a variable as used (or to remove a reported error ?)

**Comment by glenjamin** [comment]
How about something along these lines? Using the shared settings feature to ensure the rules are using the same flag on AST nodes to flag usage.

https://github.com/glenjamin/eslint/commit/d779288e694be5ea929f157ea6617276c52df791

**Comment by nzakas** [comment]
I'd like something that is a method call. Relying on people finding the variable and setting a property is too error prone.

**Comment by glenjamin** [comment]
Would that be something exposed on RuleContext? It does seem a bit too one-rule-specific to live there, but I'm not familar enough with the internals to suggest anywhere better.

**Comment by nzakas** [comment]
Yeah, probably RuleContext. We only have one documented use case for it right now, but I could easily see other rules wanting to say something was used.

**Comment by glenjamin** [comment]
Draft #2: https://github.com/glenjamin/eslint/commit/c0e979b868826ba79499724305b975cf1bd18339

**Comment by nzakas** [comment]
I think that's the idea, but we don't want people searching through the scope to find nodes. Ideally we would have something like:

```js
context.markVariableAsUsed("foo");
```

And it would search from the current scope to find that identifier and mark it as used. That way, rules wouldn't need to worry about scoping to make this work.

**Comment by glenjamin** [comment]
Here's another draft: should I be using a PR for WIP stuff? https://github.com/glenjamin/eslint/commit/aa7367c

There's currently no tests for `rule-context.js`, presumably because it's implicitly tested via all the rule tests? Once the JSX variable handling is removed from core then nothing will be using the new function - is there a way to create an anonymous rule or something similar to test this new API? If not i guess I'll have to do some similar bootstrapping to what eslint-tester does to get an AST I can test against?

**Comment by nzakas** [comment]
The functionality should live on the `eslint` object and be exposed on `RuleContext`. That's how most of the methods work. Then you can write a test using `eslint` (plenty of examples in the test file).

**Comment by glenjamin** [comment]
Ok, latest WIP is pushed to glenjamin/eslint@852be6d

I'm not really happy with my approach for testing that the `no-unused-vars` rule can be influenced by calling `markVariableAsUsed`, but I can't see another way to run a custom rule alongside a core rule.

Any suggestions?

**Comment by nzakas** [comment]
I think that's fine. Can you for a PR? I think we are close on this.
